### ISSUE TITLE
429 Error for any use of ImportJSON on coinmarkcap.com

### ISSUE THREAD

**Issue by Mark-BC** [bradjasper/ImportJSON#90]
The JSON request 
https://api.coinmarketcap.com/v2/listings/
works fine if I type it directly into a browser.

But if I put it into a Google Sheet as
=ImportJSON("https://api.coinmarketcap.com/v2/listings/","","")
then all I get is a 429 error.  It used to work.  This is a new behaviour

More specifically:
"Request failed for https://api.coinmarketcap.com/v2/listings/ returned code 429. Truncated server response: <!DOCTYPE html>  <!--[if IE 7]> <html class="no-js ie7 oldie" lang="en... (use muteHttpExceptions option to examine full response). (line 212)."

My Google Sheet has been using ImportJSON for over a year, and only recently did I discover that it has stopped working.  I was using ImportJSON v1.2.1   I have updated to v1.4.0 but that did not change anything.

**Comment by berevoescu** [comment]
I had a similar issue, starting a few days ago. I worked a bit in debug mode, with "muteHttpExceptions"   set to true and it turned out that they (CoinMarketCap) turned on limits for how many times the API can be accessed in a  given time. The details are a bit inconsistent - some say no more than 10 calls a minute, some no more than 30 - but the response shows that my IP was temporarily flagged. I stopped calling the API and after a day or two I got access again.
I am trying now to find a way to call the API only once (for all the list of currencies) and pick what I need out of the returning array, rather than calling the API once per currency.

**Comment by Mark-BC** [comment]
Thanks very much for this information.  My spreadsheet is tracking the
10-15 cryptos I care about and calling the API once per crypto (plus, OOPS,
I had 2 backup copies of that tab, so I was doing 50 or so calls whenever I
refreshed).

If you come up with a solution for downloading multiple cryptos currency
values with one call I would very much appreciate a copy of that code.
Even better would be if the code could include "convert=CAD"

Thanks!
Mark

**Comment by berevoescu** [comment]
No problem. I will share. I'm just swamped at my job with a release. I will try to do it in a couple of weeks. In the meanwhile I'd suggest my quick workaround: I temporarily use the CryptoCompare API. It is less used, but it's pretty good (I stopped using it a while ago when they had some problems).
Here it is their API entry point with an example:
https://min-api.cryptocompare.com/data/price?fsym=BTC&tsyms=BTC,USD,EUR"

And here is how I used it in my spreadsheet to get the BTC price in USD:

=ImportJSON("https://min-api.cryptocompare.com/data/price?fsym=BTC&tsyms=BTC,USD,EUR","/USD", "noHeaders, noTruncate, noInherit",dummyValueToForceUpdate)

The dummyValueToForceUpdate is not necessary. I use it with a different mechanism to automatically update after a number of minutes.

**Comment by berevoescu** [comment]
Just checked: they have CAD

https://min-api.cryptocompare.com/data/price?fsym=BTC&tsyms=BTC,USD,EUR,CAD

**Comment by Mark-BC** [comment]
Your CryptoCompare solution is perfect for me.  A single call gets all the
cryptos I want, and converted to CAD to boot.

Here is the call I am using:
=ImportJSON("
https://min-api.cryptocompare.com/data/pricemulti?fsyms=BTC,ETH,LTC,BTG,BCH,ADX,BNB,DGB,EOS,MIOTA,KMD,NEO,OMG,PAY,PLR,PPT,THETA,VEN,VERI&tsyms=CAD
",
"/BTC/CAD,/ETH/CAD,/LTC/CAD,/BTG/CAD,/BCH/CAD,/ADX/CAD,/BNB/CAD,/DGB/CAD,/EOS/CAD,/MIOTA/CAD,/KMD/CAD,/NEO/CAD,/OMG/CAD,/PAY/CAD,/PLR/CAD,/PPT/CAD,/THETA/CAD,/VEN/CAD,/VERI/CAD"
,"")
Very nice.

I'd like to know how you use the "dummyValueToForceUpdate".
My only mechanism is to alternately delete/insert a row above the
ImportJSON cell row.

Thanks very much for this pointer. I won't be using the CoinMarketCap API
any more.

Mark

**Comment by berevoescu** [comment]
Glad it works fine for you. As I said, I used this API in the past and sometimes it is less stable than I'd like. Also, they also have rate limits - they are just a bit bigger and fewer people use it (compared with CoinMarketCap) - so for now probably you won't see any rate limiting errors.
Keep in mind that we are not controlling the IP the calls come from and most likely the same IP is "shared" by many users of Google Spreadsheet. That's what it amounts to a lot of rate limiting issues.

I am checking how I can avoid that - ideally I could call a  local script that uses something like curl to get the data - that way I will have way less probability that I will go over call limits. That combined with the one-call  strategy.

On the dummyValue, it is pretty simple  and rudimentary: I add a trigger that calls a function - that function only increments a cell value. That cell value is the value I use as that dummy parameter. That means the trigger updates the dummy value every N minutes and when that happens the get prices data call is also triggered.

**Comment by andreipotorac** [comment]
@berevoescu Were you able to do this?

"I am trying now to find a way to call the API only once (for all the list of currencies) and pick what I need out of the returning array, rather than calling the API once per currency."

**Comment by berevoescu** [comment]
@andreipotorac: not yet. I am in a pretty big release and I don't have time to dedicate to this. I will try to do it in a couple of weeks or so. In the meanwhile I am working with CryptoCompare (I only need data on about 22 assets) - it is shaky (every once in a while I am getting errors from the API), but it does the job most of the time. Keep in mind that even CryptoCompare has rate limiting stops. The main problem is that many calls for the APIs come from a limited set of IPs (we don't control the IP).
As I said, the ideal scenario will be to be able to call the API locally (through something like curl) and import the data into the spreadsheet, rather than calling fetchURL). If we cannot do that, sooner or later we will collectively reach the limits, even if we call the API once for all the set of assets.

**Comment by andreipotorac** [comment]
@berevoescu Unfortunately I cannot use that since there's around 160 tokens that are listed in my sheet. I do however call the ImportJSON function just once. I guess if I allow Google to cache it (and not use a random time pegged to the link) it should return the data, although not live anymore.

I filled out CMC's form for paid access that removes the rate limit. Guess they won't have this out soon either. :(

**Comment by berevoescu** [comment]
@andreipotorac: yes, I was thinking of using the paid access too - but, as you say, unfortunately it is not clear when this will be available. Other than that we can only work with workarounds for now and no matter how much one tries, we will always depend on all the others using it - and we cannot have control over that. Another option is to transfer all to a local version and use LibreOffice - but I switched to Google Sheets because it is way more accessible from any device, it is not local etc... It's always a balance, isn't it? :-)

**Comment by andreipotorac** [comment]
@berevoescu Guess I'll give this a go again: https://jbuty.com/how-to-get-crypto-currencies-rates-and-more-in-google-sheet-1a57e571bc14  :)

**Comment by Mark-BC** [comment]
I was able to switch to CryptoCompare and only do a single call to get all
the cryptos I want, and converted to CAD as well.

Here is the new call I am using:
=ImportJSON("
https://min-api.cryptocompare.com/data/pricemulti?fsyms=BTC,ETH,LTC,BTG,BCH,ADX,BNB,DGB,EOS,MIOTA,KMD,NEO,OMG,PAY,PLR,PPT,THETA,VEN,VERI&tsyms=CAD
",
"/BTC/CAD,/ETH/CAD,/LTC/CAD,/BTG/CAD,/BCH/CAD,/ADX/CAD,/BNB/CAD,/DGB/CAD,/EOS/CAD,/MIOTA/CAD,/KMD/CAD,/NEO/CAD,/OMG/CAD,/PAY/CAD,/PLR/CAD,/PPT/CAD,/THETA/CAD,/VEN/CAD,/VERI/CAD"
,"")
Very nice.

**Comment by berevoescu** [comment]
@andreipotorac: I did try that add-on in the past, but it was pretty sketchy back then and I dropped it. I will take another look. Thanks for reminding me of it.

**Comment by berevoescu** [comment]
@andreipotorac: interesting. They did a lot of work in the meanwhile. I think I will study it in detail. It is not clear now, but since it looks like the data is collected by their side, it could be that at one point they might hit rate limits too and since the payed version of CoinmarketCap is not on yet... But maybe they have an arrangement with CMC.
For now calling CRYPTOFINANCE("COINMARKETCAP") gets everything in. It takes a while, sure. I will  look into how it gets only a subset in one call.

**Comment by andreipotorac** [comment]
@berevoescu Yeah I was afraid of the same thing, but any solution at this point - even temporary - would do.

**Comment by nepsilon** [comment]
CRYPTOFINANCE author here. I've put a [proxy service](https://cryptofinance.jbuty.com/#data) in place that fix ban from coinmarketcap (and binance). It's also immune to future API changes and existing users don't have anything to change in their sheet.

@berevoescu Let me know if you need any help. We have a Telegram support group too: https://t.me/cryptofinance_addon

**Comment by andreipotorac** [comment]
@berevoescu For me this update works well, thanks @nepsilon!

**Comment by HedonieBrazil** [comment]
I used that code, worked without coinmarketcap API.

=VALUE(SUBSTITUTE(SUBSTITUTE(INDEX(IMPORTDATA("https://min-api.cryptocompare.com/data/price?fsym=BTC&tsyms=BTC,USD,EUR"),0,2), "USD:",""), """", ""))

**Comment by GringoA1** [comment]
> 
> 
> I used that code, worked without coinmarketcap API.
> 
> =VALUE(SUBSTITUTE(SUBSTITUTE(INDEX(IMPORTDATA("[https://min-api.cryptocompare.com/data/price?fsym=BTC&tsyms=BTC,USD,EUR"),0,2](https://min-api.cryptocompare.com/data/price?fsym=BTC&tsyms=BTC,USD,EUR%22),0,2)), "USD:",""), """", ""))

If you want only USD price, why there is BTC and EUR?
